core/dispatch.c

```diff
@@ -874,6 +874,17 @@ dispatch_enter_dynamorio(dcontext_t *dcontext)
         if (get_at_syscall(dcontext))
             handle_post_system_call(dcontext);
 
+#ifdef X86
+        /* If the next basic block starts at a debug register value,
+         * we fire a single step exception before getting to the basic block. */
```

Comment:
Are we sure it should be *before*?  For a faulting instruction, a client needs to see the basic block event and to add and run its instrumentation for this instr before the instr is executed and the fault is raised.  I suppose a debug register fault is before the instr is executed so that is different.